userspace/libsinsp/metrics_collector.h

```diff
@@ -0,0 +1,246 @@
+	 * (compared to Go for example), there appear to be fewer formal client library projects available. Plus, we need to think 
+	 * about stability and long-term support before adding any new dependency.
+	 * 
+	 * The final fully qualified Prometheus metric name partially follows https://prometheus.io/docs/practices/naming/
+	 * Prepend namespace and subsystem with "_" delimiter to create a fully qualified metric name according to 
+	 * https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Opts + append unit with "_" delimiter
+	 * We do not follow the concept of base_units, but guarantee no units are mixed per unique `prom_metric_name_fully_qualified`
+	 * 
+	 * Example:
+	 * 
+	 * # HELP testns_falco_n_threads_total https://falco.org/docs/metrics/
+	 * # TYPE testns_falco_n_threads_total gauge
+	 * testns_falco_n_threads_total{raw_name="n_threads",example_key1="example1",example_key2="example2"} 12 1707281978248705000
+	 * # HELP testns_falco_memory_rss_megabytes https://falco.org/docs/metrics/
+	 * # TYPE testns_falco_memory_rss_megabytes gauge
+	 * testns_falco_memory_rss_megabytes{raw_name="memory_rss",example_key1="example1",example_key2="example2"} 350.000000 1707281978248635000
+	 * 
+	 * This method is a work in progress.
+	*/
+	std::string convert_metric_to_prom_text(metrics_v2 metric, std::string_view prom_namespace = "", std::string_view prom_subsystem = "", std::map<std::string,std::string> const_labels = {});
+
+	/*!
+	\brief Method to convert a software version like metric to the text-based Prometheus exposition format.
+	 * 
+	 * Note: Instead of using const_labels, which is a rare use case according to https://prometheus.io/docs/instrumenting/writing_exporters/#target-labels-not-static-scraped-labels, 
+	 * exposing an overload to support metrics similar to https://www.robustperception.io/exposing-the-software-version-to-prometheus/.
+	 * This approach is applicable to https://falco.org/docs/metrics/, such as Falco's "Base Fields" like falco.kernel_release and falco.version.
+	 *
+	 * Example:
+	 * 
+	 * # HELP testns_falco_kernel_release_info https://falco.org/docs/metrics/
+	 * # TYPE testns_falco_kernel_release_info untyped
+	 * testns_falco_kernel_release_info{raw_name="kernel_release",kernel_release="6.6.7-200.fc39.x86_64"} 1 1707286535681433000
+	 */
+	std::string convert_metric_to_prom_text(std::string_view metric_name, std::string_view prom_namespace = "", std::string_view prom_subsystem = "", std::map<std::string,std::string> const_labels = {});
+
+	/*!
+	\brief Method to convert memory units; tied to metrics_v2 definitions
+	*/
+	template <typename T>
+	static double convert_memory(metrics_v2_value_unit source_unit, metrics_v2_value_unit dest_unit, T val)
+	{
+		double factor = 1;
+		switch(source_unit)
+		{
+		case METRIC_VALUE_UNIT_MEMORY_BYTES:
+			factor = 1;
+			break;
+		case METRIC_VALUE_UNIT_MEMORY_KILOBYTES:
+			factor = 1024.;
```

Comment:
Yes I can change this, btw do you have more context around why base units are deemed best practice versus just saying: define one unit and don't mix units? https://github.com/falcosecurity/libs/pull/1652#issuecomment-1931452148 -> seconds would not work for the libbpf stats metrics. 

In addition, Prometheus is just one metrics consumption option. We feature internal alert rules for environments where Prometheus is not set up and we expose a config to convert memory to MB https://falco.org/docs/metrics/falco-metrics/. Wouldn't we make the interface more difficult to follow if we now say for Prometheus you cannot use this option and all memory metrics can only be in bytes?

Reply:
Another idea could be to expose a Falco config option called `convert_metrics_to_prometheus_base_unit` and then it's up to the user to decide if 200 ns is nice to look at as 2e-7 seconds etc. WDYT? 
CC @Andreagit97 if I recall correctly you had also mentioned base units.

Reply:
> why base units are deemed best practice versus just saying: define one unit and don't mix units
I found some reading material to put out there, rather than my personal opinion, so hopefully that'll be canonical enough 😄 
See:
* [the original discussion in OpenMetrics about whether to mandate that base units MUST or SHOULD be used](https://github.com/OpenObservability/OpenMetrics/issues/31)
* [the relevant section of the OpenMetrics spec that has some good consideration as to why](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#units-and-base-units)

I think for me the argument that hits home is twofold:
* I don't need to think about units. Right now all the metrics I am scraping for instance that are related to a duration are in seconds, and I don't need to wrangle magic `/ 10e3` in some part of a query to convert part of the data that I may want to compare to some other that may be in a different unit. It sounds like it would be pretty error prone
* It leaves the job of dealing with human-readable numbers to whatever representation tooling you use. That let's e.g. Grafana do the job of formatting it to something that's not e.g. "12639271 seconds" but neatly plotted as a human readable string, while Prometheus will still have the raw value ready to be math-ed on.

To be clear, I'm not suggesting you *have to* use base units. After the all the OpenMetrics spec itself recognises that that's not always feasible. It's just that *if* you can, it tends to make it easier on user then to use, aggregate and correlate that data against systems they may be scraping 🙂

Reply:
I can see that dilemma, units are hard anyways. @leogr @Andreagit97 and @jasondellaluce would you have an opinion: Should we only expose Prometheus  w/ base_units or expose several options through configs? 

Looking at just libs we mostly keep units raw (e.g. `m_vmrss_kb`) and that extended to the current metrics. Do we want to make things uniformly very Prometheus specific? What if we then want to support another standard in the future that is in slight conflict?

Another current issue is that for example libbpf stats have `ns` in the native metric name. We are now sometimes appending the unit again in the Prom text conversion, possibly even a different unit. 

The newer Metrics fields are still fairly new in Falco, we could break all metrics names, it's one option or we can have Prometheus specific patches and keep the metrics names as they are for the Falco output_rule options.

Reply:
Something about it came to my mind: shouldn't the actual units handling be left to the endpoint providing the information rather than in the core libraries ?

The role of the libraries would be to provide metrics in a consistent and simple way and then whatever is selected as metrics provider should do the conversion(s). This would be implemented at the application level so if at some point one would like to provide an alternative to Prometheus, then there's no complex "de-prometheusification" to do.

If it should stay in the library, I would suggest to change `convert_metric_to_prometheus_text` to `convert_metric_to_text`, make it virtual and then create a Prometheus specific subclass. This way it become easier to implement other providers.

Reply:
> If it should stay in the library, I would suggest to change `convert_metric_to_prometheus_text` to `convert_metric_to_text`, make it virtual and then create a Prometheus specific subclass. This way it become easier to implement other providers.

I agree with this. I'm not even sure why we should give libs the responsibility to interface with Prometheus at this stage. cc @jasondellaluce

Reply:
How about keeping it managed here in this class as we have had it many times that something lived in falco and then we decided to move it to libs. Re a more generic `convert_metric_to_text` (I agree), however instead of a subclass we could also introduce a function type for the conversion function? @sgaist @leogr any strong style preference?

@leogr would you have a stance re the units conversions in general? Would appreciate finding consensus 🙏

Reply:
I don't have a style preference.

For consensus I believe we need more people here cc @falcosecurity/libs-maintainers

Reply:
Pushed a suggestion, perhaps we can further discuss next week?
Let's definitely defer the Prometheus units conversions to a new follow up PR.